File: packages/ik_rules/lib/src/guild/skill_milestones.dart
// Guild chat announcements when a member crosses skill thresholds.

class GuildSkillMilestoneSettings {
  const GuildSkillMilestoneSettings({
    required this.enabled,
    required this.levelStart,
    required this.levelStep,
    required this.xpStartMillion,
    required this.xpStepMillion,
  });

  final bool enabled;
  final num levelStart;
  final num levelStep;
  final num xpStartMillion;
  final num xpStepMillion;

  Map<String, Object?> toJson() => <String, Object?>{
    'enabled': enabled,
    'levelStart': levelStart,
    'levelStep': levelStep,
    'xpStartMillion': xpStartMillion,
    'xpStepMillion': xpStepMillion,
  };
}

const GuildSkillMilestoneSettings defaultGuildSkillMilestoneSettings = GuildSkillMilestoneSettings(
  enabled: true,
  levelStart: 50,
  levelStep: 10,
  xpStartMillion: 125,
  xpStepMillion: 25,
);

GuildSkillMilestoneSettings normalizeGuildSkillMilestoneSettings(Object? raw) {
  final base = defaultGuildSkillMilestoneSettings;
  if (raw is! Map) return base;
  num read(String key, num fallback) {
    final value = raw[key];
    if (value is num && value.isFinite) return value.floor().clamp(1, 1 << 30);
    return fallback;
  }

  return GuildSkillMilestoneSettings(
    enabled: raw['enabled'] != false,
    levelStart: read('levelStart', base.levelStart),
    levelStep: read('levelStep', base.levelStep),
    xpStartMillion: read('xpStartMillion', base.xpStartMillion),
    xpStepMillion: read('xpStepMillion', base.xpStepMillion),
  );
}

sealed class GuildSkillMilestone {
  const GuildSkillMilestone({required this.skillId, required this.skillName});

  final String skillId;
  final String skillName;
}

class GuildSkillLevelMilestone extends GuildSkillMilestone {
  const GuildSkillLevelMilestone({
    required super.skillId,
    required super.skillName,
    required this.level,
  });

  final num level;
}

class GuildSkillXpMilestone extends GuildSkillMilestone {
  const GuildSkillXpMilestone({
    required super.skillId,
    required super.skillName,
    required this.xpMillion,
  });

  final num xpMillion;
}

String formatGuildSkillMilestone(String characterName, GuildSkillMilestone milestone) {
  final name = characterName.trim().isEmpty ? 'Adventurer' : characterName.trim();
  if (milestone is GuildSkillLevelMilestone) {
    return '$name reached ${milestone.skillName} ${milestone.level.round()}';
  }
  if (milestone is GuildSkillXpMilestone) {
    return '$name reached ${milestone.xpMillion.round()}m ${milestone.skillName} XP';
  }
  return name;
}

List<num> levelMilestonesAtOrBelow(num level, GuildSkillMilestoneSettings settings) {
  if (level < settings.levelStart) return const <num>[];
  final out = <num>[];
  for (var n = settings.levelStart; n <= level; n += settings.levelStep) {
    out.add(n);
  }
  return out;
}

List<num> xpMillionMilestonesAtOrBelow(num xp, num level, GuildSkillMilestoneSettings settings) {
  if (level < 100) return const <num>[];
  final millions = (xp / 1000000).floor();
  if (millions < settings.xpStartMillion) return const <num>[];
  final out = <num>[];
  for (var n = settings.xpStartMillion; n <= millions; n += settings.xpStepMillion) {
    out.add(n);
  }
  return out;
}

List<GuildSkillMilestone> guildSkillMilestonesCrossed({
  required String skillId,
  required String skillName,
  required num beforeLevel,
  required num beforeXp,
  required num afterLevel,
  required num afterXp,
  required GuildSkillMilestoneSettings settings,
}) {
  final normalized = normalizeGuildSkillMilestoneSettings(settings.toJson());
  if (!normalized.enabled) return const <GuildSkillMilestone>[];

  final beforeLevels = levelMilestonesAtOrBelow(beforeLevel, normalized).toSet();
  final out = <GuildSkillMilestone>[];
  for (final level in levelMilestonesAtOrBelow(afterLevel, normalized)) {
    if (!beforeLevels.contains(level)) {
      out.add(GuildSkillLevelMilestone(skillId: skillId, skillName: skillName, level: level));
    }
  }

  final beforeXpMarks = xpMillionMilestonesAtOrBelow(beforeXp, beforeLevel, normalized).toSet();
  for (final xpMillion in xpMillionMilestonesAtOrBelow(afterXp, afterLevel, normalized)) {
    if (!beforeXpMarks.contains(xpMillion)) {
      out.add(
        GuildSkillXpMilestone(skillId: skillId, skillName: skillName, xpMillion: xpMillion),
      );
    }
  }
  return out;
}

class _SkillSnap {
  const _SkillSnap({required this.skillId, required this.level, required this.xp});

  final String skillId;
  final num level;
  final num xp;
}

List<GuildSkillMilestone> guildSkillMilestonesBetweenSaves({
  required List<({String skillId, num level, num xp})> beforeSkills,
  required List<({String skillId, num level, num xp})> afterSkills,
  required String Function(String skillId) skillName,
  required GuildSkillMilestoneSettings settings,
}) {
  final normalized = normalizeGuildSkillMilestoneSettings(settings.toJson());
  if (!normalized.enabled) return const <GuildSkillMilestone>[];
  final beforeById = <String, _SkillSnap>{
    for (final row in beforeSkills)
      row.skillId: _SkillSnap(skillId: row.skillId, level: row.level, xp: row.xp),
  };
  final out = <GuildSkillMilestone>[];
  for (final after in afterSkills) {
    final before = beforeById[after.skillId];
    out.addAll(
      guildSkillMilestonesCrossed(
        skillId: after.skillId,
        skillName: skillName(after.skillId),
        beforeLevel: before?.level ?? 1,
        beforeXp: before?.xp ?? 0,
        afterLevel: after.level,
        afterXp: after.xp,
        settings: normalized,
      ),
    );
  }
  return out;
}

File: src/game/guild/skillMilestones.ts
/** Guild chat announcements when a member crosses skill thresholds. */

export interface GuildSkillMilestoneSettings {
  enabled: boolean
  /** First level that announces (default 50). */
  levelStart: number
  /** Level spacing (default 10). */
  levelStep: number
  /** First total-skill XP threshold in millions after level 100 (default 125). */
  xpStartMillion: number
  /** XP threshold spacing in millions (default 25). */
  xpStepMillion: number
}

export const DEFAULT_GUILD_SKILL_MILESTONE_SETTINGS: GuildSkillMilestoneSettings = {
  enabled: true,
  levelStart: 50,
  levelStep: 10,
  xpStartMillion: 125,
  xpStepMillion: 25,
}

const XP_MILLION = 1_000_000

export function normalizeGuildSkillMilestoneSettings(
  raw: Partial<GuildSkillMilestoneSettings> | null | undefined,
): GuildSkillMilestoneSettings {
  const base = DEFAULT_GUILD_SKILL_MILESTONE_SETTINGS
  if (!raw || typeof raw !== 'object') return { ...base }
  const levelStart = Math.max(1, Math.floor(Number(raw.levelStart) || base.levelStart))
  const levelStep = Math.max(1, Math.floor(Number(raw.levelStep) || base.levelStep))
  const xpStartMillion = Math.max(1, Math.floor(Number(raw.xpStartMillion) || base.xpStartMillion))
  const xpStepMillion = Math.max(1, Math.floor(Number(raw.xpStepMillion) || base.xpStepMillion))
  return {
    enabled: raw.enabled !== false,
    levelStart,
    levelStep,
    xpStartMillion,
    xpStepMillion,
  }
}

export type GuildSkillMilestone =
  | { kind: 'level'; skillId: string; skillName: string; level: number }
  | { kind: 'xp'; skillId: string; skillName: string; xpMillion: number }

export function formatGuildSkillMilestone(
  characterName: string,
  milestone: GuildSkillMilestone,
): string {
  const name = characterName.trim() || 'Adventurer'
  if (milestone.kind === 'level') {
    return `${name} reached ${milestone.skillName} ${milestone.level}`
  }
  return `${name} reached ${milestone.xpMillion}m ${milestone.skillName} XP`
}

/** True when [body] is a guild skill-milestone line, which already names the player. */
export function isGuildSkillMilestoneBody(body: string): boolean {
  const text = body.trim()
  return /^.+ reached .+ \d+$/.test(text) || /^.+ reached \d+m .+ XP$/.test(text)
}

/** Level thresholds at or below `level` starting from settings.levelStart. */
export function levelMilestonesAtOrBelow(
  level: number,
  settings: GuildSkillMilestoneSettings,
): number[] {
  if (level < settings.levelStart) return []
  const out: number[] = []
  for (let n = settings.levelStart; n <= level; n += settings.levelStep) {
    out.push(n)
  }
  return out
}

/** XP million thresholds at or below `xp` (only meaningful after level 100). */
export function xpMillionMilestonesAtOrBelow(
  xp: number,
  level: number,
  settings: GuildSkillMilestoneSettings,
): number[] {
  if (level < 100) return []
  const millions = Math.floor(xp / XP_MILLION)
  if (millions < settings.xpStartMillion) return []
  const out: number[] = []
  for (let n = settings.xpStartMillion; n <= millions; n += settings.xpStepMillion) {
    out.push(n)
  }
  return out
}

export function guildSkillMilestonesCrossed(args: {
  skillId: string
  skillName: string
  beforeLevel: number
  beforeXp: number
  afterLevel: number
  afterXp: number
  settings: GuildSkillMilestoneSettings
}): GuildSkillMilestone[] {
  const settings = normalizeGuildSkillMilestoneSettings(args.settings)
  if (!settings.enabled) return []
  if (args.afterLevel < args.beforeLevel && args.afterXp <= args.beforeXp) return []

  const beforeLevels = new Set(levelMilestonesAtOrBelow(args.beforeLevel, settings))
  const afterLevels = levelMilestonesAtOrBelow(args.afterLevel, settings)
  const out: GuildSkillMilestone[] = []
  for (const level of afterLevels) {
    if (!beforeLevels.has(level)) {
      out.push({
        kind: 'level',
        skillId: args.skillId,
        skillName: args.skillName,
        level,
      })
    }
  }

  const beforeXp = new Set(
    xpMillionMilestonesAtOrBelow(args.beforeXp, args.beforeLevel, settings),
  )
  const afterXp = xpMillionMilestonesAtOrBelow(args.afterXp, args.afterLevel, settings)
  for (const xpMillion of afterXp) {
    if (!beforeXp.has(xpMillion)) {
      out.push({
        kind: 'xp',
        skillId: args.skillId,
        skillName: args.skillName,
        xpMillion,
      })
    }
  }
  return out
}

export function guildSkillMilestonesBetweenSaves(args: {
  beforeSkills: Array<{ skillId: string; level: number; xp: number }>
  afterSkills: Array<{ skillId: string; level: number; xp: number }>
  skillName: (skillId: string) => string
  settings: GuildSkillMilestoneSettings
}): GuildSkillMilestone[] {
  const settings = normalizeGuildSkillMilestoneSettings(args.settings)
  if (!settings.enabled) return []
  const beforeById = new Map(args.beforeSkills.map((row) => [row.skillId, row]))
  const out: GuildSkillMilestone[] = []
  for (const after of args.afterSkills) {
    const before = beforeById.get(after.skillId)
    out.push(
      ...guildSkillMilestonesCrossed({
        skillId: after.skillId,
        skillName: args.skillName(after.skillId),
        beforeLevel: before?.level ?? 1,
        beforeXp: before?.xp ?? 0,
        afterLevel: after.level,
        afterXp: after.xp,
        settings,
      }),
    )
  }
  return out
}
